design/annotation-for-power-cycling-and-deleting-failed-nodes.md

```diff
@@ -95,35 +95,51 @@ itself, allowing it to host workloads.
 
 ## Proposal
 
-This proposal calls for a new controller which watches for the presence of a
-specific Node annotation.  If present, the controller will locate the Machine
-and BareMetalHost host objects via their annotations, and (in serial) use the
-BareMetalHost API to power off the machine, delete the Node, and then power the
-machine back on again.
-
-In order to track the state of the recovery workflow, a [new
-CRD](https://github.com/kubevirt/machine-remediation/blob/master/pkg/apis/machineremediation/v1alpha1/machineremediation_types.go)
-is proposed.
+This proposal calls for a new machine remediation controller (MRC) in CAPBM
+which watches for the presence of a `host.metal3.io/external-remediation` Machine
+annotation. If present, the controller will locate the Machine
+and BareMetalHost host objects via their annotations, and will annotate the BareMetalHost
+CR with a `reboot.metal3.io/machine-remediation`. Upon host power off, the controller
+will delete the node and then remove the annotation, to power on the host by
+utilizing [Baremetal-Operator reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/reboot-interface.md).
+
+After restart/crash of the new controller it will check for
+`reboot.metal3.io/machine-remediation` annotation on all hosts,
+and continue remediation process, to make sure that such host is
+not left in powered off state.
+
+## Assumptions
+
+MHC checks that each machine has a node, and if it doesn't after some timeout,
```

Comment:
Does the MHC will be modifying CAPI level Machine object? Or should it be BareMetalMachine object?
Because, what I understood first is that, MHC will watch for Machine object, if Machine meets unhealthiness condition, then MHC will set the `host.metal3.io/external-remediation` annotation on corresponding BareMetalMachine object, meaning that MRC will be also watching for BareMetalMachine and if `host.metal3.io/external-remediation` annotation is present, then annotate corresponding BareMetalHost object with `reboot.metal3.io/machine-remediation`.
Could you please confirm, if I'm right that - MHC will watch for CAPI level Machine object but will annotate  the CAPM3 level BareMetalMachine object not the Machine object with `host.metal3.io/external-remediation`?
/cc @jan-est

Reply:
MHC doesn't modify BareMetalHost. If a Machine is unhealthy, it will annotate the **Machine** with `host.metal3.io/external-remediation`. Now MRC kicks in (watches both Machine and BareMetalHost)

Reply:
@dhellmann  
We are actually quite interested in having MRC & MHC on upstream. During the last community meeting I was asking about how problematic would it be to backport MRC code from openshift/capbm (once it is there) to Metal3. But, as you also mentioned, the API versions are different, and there is no CRD. So, I'm actually starting to write MRC in parallel and wanted to make sure that writing it in parallel is better rather than waiting for the code first to be landed on openshift/capbm and then port to upstream?

Reply:
Oh, it seems a shame to duplicate the effort of building it.

Is the only difference the type of resource the controller has to watch or update?

Reply:
Wait a second... there are `BareMetalMachine`s now too?
Can someone point me to the design doc where they were introduced?

Reply:
> Wait a second... there are `BareMetalMachine`s now too?
> Can someone point me to the design doc where they were introduced?

Maybe these pictures will help you understand the current design:
[v1alpha3](https://github.com/Nordix/metal3-docs/blob/v1a3/crs-feruz/images/v1alpha3_crs.svg)
[v1alpha2](https://github.com/metal3-io/metal3-docs/blob/master/crs_reference.md)

Reply:
> @fmuyassarov Can you clarify whether
> 
> > I'm actually starting to write MRC in parallel
> 
> means you're working on the backport already, or starting a completely new implementation?

I was just trying out backporting option while experimenting with how current implementation of MRC would work in the context of CAPM3, but didn’t start full implementation.

Reply:
> I was just trying out backporting option while experimenting with how current implementation of MRC would work in the context of CAPM3, but didn’t start full implementation.

Ah, very cool.

Regarding the diagrams, they certainly help explain what but why. Is this pattern of every object having a provider specific shadow mandated by the cluster API?

Reply:
Are you referring to the relationship/mapping between CAPI API version and Provider API version?

Reply:
I mean why Machines and Clusters have BareMetalMachines and BareMetalClusters

Reply:
Because this is the way of representing those Cluster and Machine objects in infrastructure provider level. Basically, each infra provider has their own objects (INFRAXCluster, INFRAXMachine...so on) which are eventually linked to Cluster API objects (Machines & Clusters...son on). See [AWS example](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/982d2200ded3123e3803cd0d3c9197f4af5380d0/templates/cluster-template.yaml#L12).

Reply:
My 2c: it makes sense to me to keep the annotation on Machine (not BareMetalMachine) because the remediation system shouldn't need to know details about what type of platform it is running on. (Ultimately, we would hope to make this part of the Cluster API and supported by all providers.)

Reply:
Yep, we're working on having some of this upstream